File: android-client/app/src/main/java/com/anrixa/voiceledger/network/ApiClient.kt
package com.anrixa.voiceledger.network

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.ResponseBody
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.Multipart
import retrofit2.http.POST
import retrofit2.http.Part
import retrofit2.http.Query
import retrofit2.http.Streaming
import java.io.File
import java.util.concurrent.TimeUnit

// ---- request/response bodies (mirrors app/models.py on the server) ----

data class PairStartResponse(val pairing_code: String, val expires_in: Int)
data class PairCompleteRequest(val pairing_code: String, val label: String, val public_key_pem: String)
data class PairCompleteResponse(val device_id: String)

data class SttResponse(val transcript: String)

data class TransactionInBody(
    val device_id: String,
    val payload_json: String,
    val signature_b64: String,
)

data class TransactionOut(
    val id: String,
    val counterparty_name: String,
    val place: String?,
    val amount_paid: Double?,
    val amount_paid_currency: String?,
    val amount_due: Double?,
    val amount_due_currency: String?,
    val due_date: String?,
    val transaction_date: String?,
    val created_at: String,
)

interface VoiceLedgerApi {
    @POST("pair/start")
    suspend fun pairStart(): PairStartResponse

    @POST("pair/complete")
    suspend fun pairComplete(@Body req: PairCompleteRequest): PairCompleteResponse

    @Multipart
    @POST("stt")
    suspend fun speechToText(@Part audio: MultipartBody.Part): SttResponse

    @POST("nlu/guided-field")
    suspend fun guidedField(
        @Query("field") field: String,
        @Query("transcript") transcript: String,
    ): Map<String, Any?>

    @Streaming
    @GET("tts")
    suspend fun textToSpeech(@Query("text") text: String): ResponseBody

    @POST("transactions")
    suspend fun submitTransaction(@Body body: TransactionInBody): TransactionOut

    @GET("transactions")
    suspend fun listTransactions(): List<TransactionOut>
}

/**
 * `baseUrl` is the LAN address entered (or discovered) in Settings —
 * e.g. "http://192.168.1.42:8420/". Rebuild this whenever that
 * address changes; there's no DNS to fall back on, it's just an IP.
 */
object ApiClientFactory {
    fun create(baseUrl: String): VoiceLedgerApi {
        val client = OkHttpClient.Builder()
            .connectTimeout(5, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS) // STT can take a few seconds on CPU
            .build()

        return Retrofit.Builder()
            .baseUrl(baseUrl)
            .client(client)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(VoiceLedgerApi::class.java)
    }
}

fun audioFilePart(file: File): MultipartBody.Part {
    val body = file.asRequestBody("audio/wav".toMediaType())
    return MultipartBody.Part.createFormData("audio", file.name, body)
}

private fun File.asRequestBody(mediaType: okhttp3.MediaType) =
    this.readBytes().toRequestBody(mediaType)

File: pc-client/api_client.py
"""
Talks to the same server the Android app talks to - see
server/app/main.py. Kept as plain functions (not a class) since a
desktop settings screen is the only thing that needs a base URL, and
it can just be threaded through.
"""
from pathlib import Path
from typing import Any

import requests


class ApiError(Exception):
    pass


def _url(base_url: str, path: str) -> str:
    return base_url.rstrip("/") + "/" + path.lstrip("/")


def pair_start(base_url: str) -> dict:
    resp = requests.post(_url(base_url, "/pair/start"), timeout=10)
    _raise_for_status(resp)
    return resp.json()


def pair_complete(base_url: str, pairing_code: str, label: str, public_key_pem: str) -> dict:
    resp = requests.post(
        _url(base_url, "/pair/complete"),
        json={"pairing_code": pairing_code, "label": label, "public_key_pem": public_key_pem},
        timeout=10,
    )
    _raise_for_status(resp)
    return resp.json()


def speech_to_text(base_url: str, wav_path: Path) -> str:
    with open(wav_path, "rb") as f:
        resp = requests.post(
            _url(base_url, "/stt"), files={"audio": (wav_path.name, f, "audio/wav")}, timeout=60
        )
    _raise_for_status(resp)
    return resp.json()["transcript"]


def guided_field(base_url: str, field: str, transcript: str) -> dict:
    resp = requests.post(
        _url(base_url, "/nlu/guided-field"),
        params={"field": field, "transcript": transcript},
        timeout=10,
    )
    _raise_for_status(resp)
    return resp.json()


def submit_transaction(base_url: str, device_id: str, payload_json: str, signature_b64: str) -> dict:
    resp = requests.post(
        _url(base_url, "/transactions"),
        json={"device_id": device_id, "payload_json": payload_json, "signature_b64": signature_b64},
        timeout=10,
    )
    _raise_for_status(resp)
    return resp.json()


def list_transactions(base_url: str) -> list[dict]:
    resp = requests.get(_url(base_url, "/transactions"), timeout=10)
    _raise_for_status(resp)
    return resp.json()


def fetch_tts(base_url: str, text: str) -> bytes:
    resp = requests.get(_url(base_url, "/tts"), params={"text": text}, timeout=30)
    _raise_for_status(resp)
    return resp.content


def _raise_for_status(resp: requests.Response) -> None:
    if not resp.ok:
        try:
            detail = resp.json().get("detail", resp.text)
        except Exception:  # noqa: BLE001
            detail = resp.text
        raise ApiError(f"{resp.status_code}: {detail}")
